crates/binjs_es6/src/scopes.rs

```diff
@@ -55,16 +53,23 @@ impl AnnotationVisitor {
     }
 
     fn push_free_names(&mut self) {
-        self.free_names_in_function_stack.push(HashSet::new());
+        self.free_names_in_block_stack.push(HashMap::new());
     }
-    fn pop_free_names(&mut self, bindings: &[&HashSet<String>]) {
-        let mut free_names_in_current_function = self.free_names_in_function_stack.pop().unwrap();
-        for name in free_names_in_current_function.drain() {
+    fn pop_free_names(&mut self, bindings: &[&HashSet<String>], function_scope: bool) {
+        let mut free_names_in_current_block = self.free_names_in_block_stack.pop().unwrap();
+        for (name, old_cross_function) in free_names_in_current_block.drain() {
             let is_bound = bindings.iter()
                 .find(|container| container.contains(&name))
                 .is_some();
             if !is_bound {
-                self.free_names_in_nested_functions.insert(name);
+                let parent_free = self.free_names_in_block_stack.last_mut();
+                if parent_free.is_some() {
+                    let mut unwrapped = parent_free.unwrap();
+                    if unwrapped.get(&name) == Some(&true) {
```

Comment:
Nit: Here, you should inspect a value, rather than create a new one for comparison purposes.

```rust
if let Some(&true) = unwrapped.get(&name) {
  // ...
}
```

Except in this case, it looks like you want to update the entry, so it's rather something along the lines of

```rust
let new_is_cross_function = old_cross_function || function_scope;

unwrapped.entry(&name)
  // Modify the entry if it is already present.
  .and_modify(new_is_cross_function)
  // Or insert a new value.
  .or_insert(new_is_cross_function);
```